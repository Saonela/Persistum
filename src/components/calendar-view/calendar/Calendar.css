
.calendar {
    padding: var(--margin-m) var(--margin-l);
}

.calendar__month:not(:last-child) {
    margin-bottom: var(--margin-l);
}

.calendar__days {
    display: grid;
    justify-content: center;
    grid-template-columns: repeat(7, var(--width-calendar-card));
    grid-column-gap: var(--margin-s);
    grid-row-gap: var(--margin-s);
    row-gap: var(--margin-s);
    direction: rtl;
}

.calendar-year-label {
    margin-bottom: calc(var(--margin-s) * -1);
    font-size: var(--font-size-xl);
    text-align: left;
}

.calendar-month-label {
    margin: var(--margin-s) 0 var(--margin);
    font-size: var(--font-size-l);
    text-align: left;
}

.calendar-month-label__year {
    font-size: var(--font-size);
    margin-left: var(--margin-xs);
}

.list-view .calendar__days {
    grid-template-columns: repeat(1, 1fr);
}

@media only screen and (max-width: 1300px) {
    .calendar__days {
        grid-template-columns: repeat(6, 1fr);
    }
}

@media only screen and (max-width: 1000px) {
    .calendar__days {
        grid-template-columns: repeat(5, 1fr);
    }
}


@media only screen and (max-width: 850px) {
    .calendar__days {
        grid-template-columns: repeat(4, 1fr);
    }
}

@media only screen and (max-width: 660px) {
    .calendar__days {
        grid-template-columns: repeat(3, 1fr);
    }
}

@media only screen and (max-width: 525px) {
    .calendar__days {
        grid-template-columns: repeat(2, 1fr);
    }
}
