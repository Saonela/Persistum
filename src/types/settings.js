
export const CALENDAR_DISPLAY_TYPE = {
    GRID: 'grid',
    LIST: 'list'
};
