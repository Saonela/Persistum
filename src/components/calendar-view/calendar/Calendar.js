import React from "react";
import './Calendar.css'
import moment from "moment";
import WeekdaysHeader from "../weekdays-header/WeekdaysHeader";
import {CALENDAR_DISPLAY_TYPE} from "../../../types/settings";
import CalendarCell from "../calendar-cell/CalendarCell";

function Calendar({activitiesMap, dataLog, displayType}) {

    const renderDayPlaceholder = (timestamp) => {
        const placeholders = [];
        const weekday = moment(timestamp, 'YYYY-MM-DD').day();
        const limit = 7 - (weekday === 0 ? 7 : weekday); // moment.js sunday = 0
        for (let i = 0; i < limit; i++) {
            placeholders.push(<div key={i} className="calendar__days-placeholder" data-testid="day-placeholder"/>);
        }
        return placeholders;
    };

    const mapIdToActivity = (activityIds) => {
        return activityIds.map(id => activitiesMap[id]).filter(activity => activity);
    };

    return (
        <React.Fragment>
            <div className={`calendar ${displayType === CALENDAR_DISPLAY_TYPE.LIST && 'list-view'}`}>
                {dataLog.map((yearData, i) =>
                    <div className="calendar__year" key={yearData.year} data-testid={'year-' + yearData.year}>
                        <div className="calendar__year-label">{yearData.year}</div>
                        {yearData.data.map((monthData, i) =>
                            <div className="calendar__month" key={monthData.month} data-testid={'month-' + monthData.month}>
                                <div className="calendar__month-label">{moment().month(monthData.month - 1).format("MMMM")}</div>
                                {displayType !== CALENDAR_DISPLAY_TYPE.LIST && <WeekdaysHeader/>}
                                <div className="calendar__days">
                                    {renderDayPlaceholder(monthData.data[0].timestamp)}
                                    {monthData.data.map((entryData) =>
                                        <CalendarCell className="calendar__cell"
                                                      displayType={displayType}
                                                      key={entryData.timestamp}
                                                      timestamp={entryData.timestamp}
                                                      activities={mapIdToActivity(entryData.activities)}/>
                                    )}
                                </div>
                            </div>
                        )}
                    </div>
                )}
            </div>
        </React.Fragment>
    )
}

export default Calendar;
