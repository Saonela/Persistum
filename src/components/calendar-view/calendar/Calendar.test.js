import React from "react";
import Calendar from "./Calendar";
import {render, screen} from "@testing-library/react";
import {within} from "@testing-library/dom";
import {CALENDAR_DISPLAY_TYPE} from "../../../types/settings";

describe('Calendar', () => {

    const dataLog = [
        {
            year: '2020',
            data: [
                {
                    month: '11',
                    data: [
                        {
                            timestamp: '2020-11-31',
                            activities: []
                        },
                        {
                            timestamp: '2020-11-10',
                            activities: [999]
                        }
                    ]
                },
                {
                    month: '4',
                    data: [
                        {
                            timestamp: '2020-04-31',
                            activities: []
                        },
                        {
                            timestamp: '2020-04-15',
                            activities: [999]
                        }
                    ]
                }
            ]
        },
        {
            year: '2010',
            data: [
                {
                    month: '11',
                    data: [
                        {
                            timestamp: '2010-11-30',
                            activities: []
                        },
                        {
                            timestamp: '2010-11-02',
                            activities: [123456]
                        }
                    ],
                }
            ]
        }
    ];

    const smallDataLog = [{
        year: '2010',
        data: [
            {
                month: '11',
                data: [
                    {
                        timestamp: '2010-11-30',
                        activities: []
                    },
                    {
                        timestamp: '2010-11-02',
                        activities: [123456]
                    }
                ],
            }
        ]
    }];

    const activitiesMap = {
        999: {id: 999, name: 'b', style: {}},
        123456: {id: 123456, name: 'a', style: {}}
    };

    it('should display log cells', () => {
        render(<Calendar activitiesMap={activitiesMap} dataLog={dataLog} displayType={CALENDAR_DISPLAY_TYPE.GRID}/>);
        expect(screen.getAllByRole('img', {name: 'Activity color'}).length).toBe(3);
    });

    it('should add placeholder tiles by last months day weekday', () => {
        render(<Calendar activitiesMap={activitiesMap} dataLog={dataLog} displayType={CALENDAR_DISPLAY_TYPE.GRID}/>);
        let year = screen.queryByTestId('year-2010');
        let yearTestUtils = within(year)
        let month = yearTestUtils.queryByTestId('month-11');
        let monthTestUtils = within(month)
        expect(monthTestUtils.queryAllByTestId('day-placeholder').length).toBe(5);

        year = screen.queryByTestId('year-2020');
        yearTestUtils = within(year)
        month = yearTestUtils.queryByTestId('month-4');
        monthTestUtils = within(month)
        expect(monthTestUtils.queryAllByTestId('day-placeholder').length).toBe(0);
    });

    it('display weekdays header if grid view', () => {
        render(<Calendar activitiesMap={activitiesMap} dataLog={smallDataLog} displayType={CALENDAR_DISPLAY_TYPE.GRID}/>);
        screen.getByText('Monday');
        screen.getByText('Wednesday');
        screen.getByText('Friday');
        screen.getByText('Sunday');
    });

    it('display hide weekdays header if list view', () => {
        render(<Calendar activitiesMap={activitiesMap} dataLog={smallDataLog} displayType={CALENDAR_DISPLAY_TYPE.LIST}/>);
        expect(screen.queryByText('Monday')).not.toBeInTheDocument();
        expect(screen.queryByText('Wednesday')).not.toBeInTheDocument();
        expect(screen.queryByText('Friday')).not.toBeInTheDocument();
        expect(screen.queryByText('Sunday')).not.toBeInTheDocument();
    });

    it('should match snapshot', () => {
        const {asFragment} = render(<Calendar activitiesMap={activitiesMap} dataLog={smallDataLog} displayType={CALENDAR_DISPLAY_TYPE.LIST}/>);
        expect(asFragment()).toMatchSnapshot();
    });

});
