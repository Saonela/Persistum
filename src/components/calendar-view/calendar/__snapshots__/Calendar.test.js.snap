// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calendar should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="calendar list-view"
  >
    <div
      class="calendar__year"
      data-testid="year-2010"
    >
      <div
        class="calendar__year-label"
      >
        2010
      </div>
      <div
        class="calendar__month"
        data-testid="month-11"
      >
        <div
          class="calendar__month-label"
        >
          November
        </div>
        <div
          class="calendar__days"
        >
          <div
            class="calendar__days-placeholder"
            data-testid="day-placeholder"
          />
          <div
            class="calendar__days-placeholder"
            data-testid="day-placeholder"
          />
          <div
            class="calendar__days-placeholder"
            data-testid="day-placeholder"
          />
          <div
            class="calendar__days-placeholder"
            data-testid="day-placeholder"
          />
          <div
            class="calendar__days-placeholder"
            data-testid="day-placeholder"
          />
          <div
            class="calendar-cell 
                         
                         list-view"
          >
            <div
              class="calendar-cell__timestamp"
            >
              <span>
                30
              </span>
              <span
                class="calendar-cell__weekday "
              >
                Tu
              </span>
            </div>
            <div
              class="calendar-cell__activities"
            />
          </div>
          <div
            class="calendar-cell 
                         
                         list-view"
          >
            <div
              class="calendar-cell__timestamp"
            >
              <span>
                02
              </span>
              <span
                class="calendar-cell__weekday "
              >
                Tu
              </span>
            </div>
            <div
              class="calendar-cell__activities"
            >
              <div
                class="log-activity"
              >
                <div
                  aria-label="Activity color"
                  class="activity-color"
                  role="img"
                />
                <p
                  class="log-activity__name"
                >
                  a
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
